src/platform/packages/shared/kbn-esql-validation-autocomplete/src/validation/helpers.ts

```diff
@@ -96,3 +38,49 @@ export function getMaxMinNumberOfParams(definition: FunctionDefinition) {
  */
 export const getEnrichCommands = (commands: ESQLCommand[]): ESQLCommand[] =>
   Walker.matchAll(commands, { type: 'command', name: 'enrich' }) as ESQLCommand[];
+
+/**
+ * Returns a list of subqueries to validate
+ * @param rootCommands
+ */
+export function getSubqueriesToValidate(rootCommands: ESQLCommand[]) {
+  const subsequences = [];
+  const expandedCommands = expandEvals(rootCommands);
+  for (let i = 0; i < expandedCommands.length; i++) {
+    const command = expandedCommands[i];
+
+    // every command within FORK's branches is its own subquery to be validated
+    if (command.name.toLowerCase() === 'fork') {
+      const branchSubqueries = getForkBranchSubqueries(command as ESQLCommand<'fork'>);
+      for (const subquery of branchSubqueries) {
+        subsequences.push([...expandedCommands.slice(0, i), ...subquery]);
+      }
+    }
+
+    subsequences.push(expandedCommands.slice(0, i + 1));
+  }
+
+  return subsequences.map((subsequence) => Builder.expression.query(subsequence));
+}
+
+/**
+ * Expands a FORK command into flat subqueries for each command in each branch.
+ *
+ * E.g. FORK (EVAL 1 | LIMIT 10) (RENAME foo AS bar | DROP lolz)
+ *
+ * becomes [`EVAL 1`, `EVAL 1 | LIMIT 10`, `RENAME foo AS bar`, `RENAME foo AS bar | DROP lolz`]
+ *
+ * @param command a FORK command
+ * @returns an array of expanded subqueries
+ */
+function getForkBranchSubqueries(command: ESQLCommand<'fork'>): ESQLCommand[][] {
```

Comment:
Same here, I wonder if it would be easy to move on the command (not in this PR, just brainstorming)

Reply:
I also looked into deduplicating this with the logic in `getQueryForFields`. They sort of overlap since they both involve flattening FORK branches. But, it didn't really work out for me for some reason.

Reply:
yeah exactly ....